CAF-2750: Architecture documentation re-write for Audit 3.0.0 release

CJComac opened the pull request:
Architecture re-write for Audit library direct-to-elasticsearch implementation

Commit:
Architecture documentation re-write for Audit 3.0.0 release

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-Auditing is built on Apache Kafka for the messaging of the audit events and HPE Vertica for the storage of the audit events. Both of these technologies offer high availability, throughput, scalability, and performance to the overall solution. Additionally, Vertica offers strong data analytics capabilities and comes with a pre-built Kafka integration, which can continually load data from Kafka.
-
-A Mesos/Marathon environment runs the Audit Management web service and Kafka-Vertica Scheduler components, and provides redundancy for these services.
+Auditing is built on Elasticsearch for the messaging and storage of the audit events. Elasticsearch offers high availability, throughput, scalability, and performance to the overall solution. Additionally, Elasticsearch offers strong data analytics capabilities.
```
Chris you have taken the features of Apache Kakfa and Vertica here from original document and used for Elasticsearch puproses. They may still in fact be the same but there may be additional features that Elasticsearch provides. These will need added anyhow to the Features section as well in a separate ticket.

CJComac replied:
I've adjusted the second sentence in this paragraph in the latest commit: "Additionally, Elasticsearch is accessible via RESTful APIs, it offers strong data analytics and monitoring capabilities."

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-5. The audited application makes calls to the generated client-side library to send tenant audit events to Kafka messaging.
-	1. Kafka receives messages from the client-side library and stores them in per application, per tenant topics.
-	2. Kafka-Vertica Scheduler listens to the topics that are registered with it and streams them into the tenant's application table in Vertica.
+1. Setting up your application for Auditing requires defining an audit event definition XML file. The file is used for generation of the client-side audit library.
```
"... the file is used for *the* generation of the ..."

CJComac replied:
Adjusted in the latest commit

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-A Mesos/Marathon environment runs the Audit Management web service and Kafka-Vertica Scheduler components, and provides redundancy for these services.
+Auditing is built on Elasticsearch for the messaging and storage of the audit events. Elasticsearch offers high availability, throughput, scalability, and performance to the overall solution. Additionally, Elasticsearch offers strong data analytics capabilities.
 
 ### Audit Management Component Architecture
```
This is no longer Management. Replace with Audit Service Component Architecture.

CJComac replied:
Changed in the latest commit

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-	2. Kafka-Vertica Scheduler listens to the topics that are registered with it and streams them into the tenant's application table in Vertica.
+1. Setting up your application for Auditing requires defining an audit event definition XML file. The file is used for generation of the client-side audit library.
+2. Using the `caf-audit-maven-plugin`, the client-side Java library is generated from the audit event definition XML file.
+3. The audited application makes calls to the generated client-side library to send tenant audit events to Elasticsearch which stores them in per tenant indices.
```
To me, this reads a single tenant's audit events ends up in multiple tenant indexes. Just make it clear that audit events belonging to a tenant are stored in a single ES tenant index.

glucy replied:
Point 3. is confusing.

"An application being audited uses the generated client-side library to create and send audit events to Elasticsearch. 
 A new index is created per tenant in the Elasticsearch.  Therefore the audit events are stored per tenant within Elasticsearch."

Something like the above.  Although I am not sure that is the style you are looking for or if it covers enough detail.

CJComac replied:
Adjusted to point 3 to: "The audited application makes calls to the generated client-side library to send audit events to Elasticsearch. An Audit event is stored in the index belonging to the tenant that made the call. In other words, audit events are stored per tenant within Elasticsearch."

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-The figure shows the `account_1` schema with a `kafka_rej` table and columns for rejected audit event data.
+### Elasticsearch Indices and Type Mappings
+
+On a tenant's first call of the Audit library, an index is created for the them. A tenant's Elasticsearch index, type and document's ID (`index/type/_id`) is seen as `audit_tenant_<tenantId>/cafAuditEvent/<applicationAuditEvent>`.
```
A tenant does not call the Audit Library.
Also in ES an indexed document is uniquely identified by index/type/_id where _id (as our solution stands right now) is an auto-generated value. This value will never be "<applicationAuditEvent>" so I'm a bit confused with this sentence?.

glucy replied:
"... an index is create for the [tenant]." - Avoid them, it, etc.  Always be explicit in documentation.

CJComac replied:
- Would it be better worded that a tenant application calls the Audit library? I'm trying to describe to the reader how and when an index is created. In my mind an index is created when a new tenantId comes along. How can I better describe this?
- We don't know what the id for the audit event is but the purpose of `audit_tenant_<tenantId>/cafAuditEvent/<applicationAuditEvent>` is to describe the "index meta-fields" (or index schema) of a tenant index.

In the latest commit:
- I have removed "id" and replaced with "doc", the sentence reads: "A tenant's Elasticsearch index, type and document meta-field identifiers (`index/type/doc`) are seen as `audit_tenant_<tenantId>/cafAuditEvent/applicationAuditEvent`.".
- I have replaced "them" with "tenant".

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
+            "dynamic_templates": [
+              {
+                "CAFAuditString": {
+                  "match": "*_CAStr",
```
Chris, can you double check with Dermot over these suffixes to make sure they are def what we all agreed? Also the short suffix needs changes to something else as Dermot did not like this. Agree with him on this.

CJComac replied:
Ok will do. **TODO: Verify type mapping fields with Dermot**

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
+                }
+              }
+            ],
+            "properties": {
```
Not sure why, but I would have preferred the fixed fields list first in this mapping with the custom ones after.

CJComac replied:
On query, this is how Elasticsearch returns type mappings.

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
+      }
+    }
+
+The above JSON, returned from Elasticsearch, shows us the fixed and custom field type mappings for a tenant's index. An audit event's information is stored in fixed fields and the event's parameters are mapped to their appropriate types based on their field name suffixes added by the Audit library.
```
Should be " ... are stored ...". Also why pick out fixed fields here?

glucy replied:
This:  The above JSON, returned from Elasticsearch, shows us the fixed and custom field type mappings for a tenant's index.
Can be written:  The above JSON, returned from Elasticsearch, illustrates the fixed and custom field type mappings for an index.

glucy replied:
This:  An audit event's information is stored in fixed fields and the event's parameters are mapped to their appropriate types based on their field name suffixes added by the Audit library.
Can be written:  Audit event information is stored in fixed fields and the event parameters are mapped to appropriate types based on field name suffixes added by the Audit library.

Notice the difference in style.

CJComac replied:
This paragraph has been adjusted to "The above JSON, returned from Elasticsearch, illustrates the field type mappings for a tenant's index. Audit event information is stored in fixed fields and the event parameters are mapped to appropriate types based on field name suffixes added by the Audit library." in the latest commit

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
+
+The above JSON, returned from Elasticsearch, shows us the fixed and custom field type mappings for a tenant's index. An audit event's information is stored in fixed fields and the event's parameters are mapped to their appropriate types based on their field name suffixes added by the Audit library.
+
+A tenant application's audit events sent from the client-side library are added that tenant's index.
```
" ... are added to that tenant's index "

glucy replied:
"An application's audit events for a tenant are sent from the client-side library to Elasticsearch and added to the index created for the tenant."

Or such like ...

CJComac replied:
Adjusted to "A tenant application's audit events are sent from the client-side library to Elasticsearch and added to the index created for the tenant." in the latest commit.

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
+      }
+    }
+
+The above JSON, returned from Elasticsearch, shows us all of the audit events belonging to a tenant.
```
Specify exactly which tenant is used in the example. I'd also review the choice of test data for this example.

glucy replied:
The above JSON, returned from Elasticsearch, displays all of the audit events belonging to "audit_tenant_1"

CJComac replied:
Adjusted to "The above JSON, returned from Elasticsearch, shows us all of the audit events belonging to `audit_tenant_1`." 

The events used here are based on the example audit event definition file in the current getting started guide. **TODO**: I will review if this is OK with Dermot.

glucy commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-On the server-side, with the Audit Management web service API, the application's audit event definition file is used to create an audit event schema for the application within Vertica. The web service API can then register a tenant with one or more applications. It creates application audit event tables for the tenant and associates application tenant topics with the Kafka-Vertica Scheduler.
-
-Apache Kafka receives Audit events for an application's tenant from the client-side library and partitions them into per application, per tenant topics. The Kafka-Vertica Scheduler listens to these topics and streams the events to the tenant's application audit table in Vertica.
+Elasticsearch receives application audit events for a tenant from the client-side library and adds the application audit event to the tenant's index.
```
Elasticsearch receives [CAF audit events] for a tenant from the client-side library and adds the application audit event to the tenant's index.

My change is inside [ ]

CJComac replied:
Adjusted in the latest commit.

glucy commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
 # Architecture
 
-Auditing logs audit events on a per application, per tenant basis.  The Audit Management web service facilitates the addition of new applications and new tenants.
+The Audit library logs audit events on a per tenant, per application basis and facilitates the addition of new tenants and new applications.
```
Not sure it is exactly "per tenant", "per application".   May need to revisit this sentence.

CJComac replied:
Adjusted to just "per tenant" in the latest commit.

glucy commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,224 @@ title: Architecture
-Auditing is built on Apache Kafka for the messaging of the audit events and HPE Vertica for the storage of the audit events. Both of these technologies offer high availability, throughput, scalability, and performance to the overall solution. Additionally, Vertica offers strong data analytics capabilities and comes with a pre-built Kafka integration, which can continually load data from Kafka.
-
-A Mesos/Marathon environment runs the Audit Management web service and Kafka-Vertica Scheduler components, and provides redundancy for these services.
+Auditing is built on Elasticsearch for the messaging and storage of the audit events. Elasticsearch offers high availability, throughput, scalability, and performance to the overall solution. Additionally, Elasticsearch offers strong data analytics capabilities.
```
I agree with Connor.  What you have wrote isn't necessarily wrong but we should tailor it a little more towards Elasticsearch:

You don't need everything but this will give you some ideas:  http://whatis.techtarget.com/definition/ElasticSearch

CJComac replied:
I've adjusted the second sentence in this paragraph in the latest commit: "Additionally, Elasticsearch is accessible via RESTful APIs, it offers strong data analytics and monitoring capabilities."

CJComac reviewed (commented):
Feedback considered and changes made in the latest commit.

Commit:
Made Greg and Connor's feedback changes.

Commit:
Adjusted type mappings JSON fixed fields' datatypes and tenant index naming to match those revised during CAF-2613.

glucy commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,216 @@ title: Architecture
+### Elasticsearch Indices and Type Mappings
+
+On a tenant's first call of the Audit library, an index is created for the tenant. A tenant's Elasticsearch index, type and document meta-field identifiers (`index/type/doc`) are seen as `audit_tenant_<tenantId>/cafAuditEvent/applicationAuditEvent`.
+
```
`audit_tenant_<tenantId>/cafAuditEvent/applicationAuditEvent`

Should this not now be updated to:  `<tenantId>_audit/cafAuditEvent/applicationAuditEvent`

?

CJComac replied:
Adjusted in the latest commit

glucy commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,216 @@ title: Architecture
+      }
+    }
+
+The above JSON, returned from Elasticsearch, illustrates the field type mappings for a tenant's index. Audit event information is stored in fixed fields and the event parameters are mapped to appropriate types based on field name suffixes added by the Audit library.
```
Can this be better worded? "The above JSON, returned from Elasticsearch, illustrates the field type mappings for a tenant's index."

Say:

"The above JSON, returned from Elasticsearch, illustrates the field type mappings for an index belonging to a tenant with <tenantId> == 1 i.e. 1_audit."

CJComac replied:
Reworded to "The above JSON, returned from Elasticsearch, illustrates the field type mappings for an index belonging to a tenant who's ID is `1`. " in the latest commit.

glucy reviewed (commented):
Once these changes are made this looks good to me.

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,216 @@ title: Architecture
 # Architecture
 
-Auditing logs audit events on a per application, per tenant basis.  The Audit Management web service facilitates the addition of new applications and new tenants.
+The Audit library logs audit events on a per tenant basis and facilitates the addition of new tenants and new applications.
```
The audit library does not facilitate the addition of new tenants and applications. The audit library is designed to index tenant/application audit event metadata into Elasticsearch. In our current audit solution, we register tentants and applications but I don't think it should be worded like this for our Elasticsearch implementation.

CJComac replied:
Reworded to "The Audit library logs audit event metadata, belonging to tenant applications, into Elasticsearch." in the latest commit.

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,216 @@ title: Architecture
-	2. Kafka-Vertica Scheduler listens to the topics that are registered with it and streams them into the tenant's application table in Vertica.
+1. Setting up your application for Auditing requires defining an audit event definition XML file. The file is used for the generation of the client-side audit library.
+2. Using the `caf-audit-maven-plugin`, the client-side Java library is generated from the audit event definition XML file.
+3. The audited application makes calls to the generated client-side library to send audit events to Elasticsearch. An Audit event is stored in the index belonging to the tenant that made the call. In other words, audit events are stored per tenant within Elasticsearch.
```
I'm not sure a tenant makes a call here. The audit application will utilise the audit library to index tenant related audit metadata but the tenant itself does not make the call.

CJComac replied:
I reworded the third bullet point to "The audited application makes calls to the generated client-side library to send audit events to Elasticsearch. An Audit event is stored in the tenant index belonging to the application that made the call."

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,216 @@ title: Architecture
-The figure shows the `account_1` schema with a `kafka_rej` table and columns for rejected audit event data.
+### Elasticsearch Indices and Type Mappings
+
+On a tenant's first call of the Audit library, an index is created for the tenant. A tenant's Elasticsearch index, type and document meta-field identifiers (`index/type/doc`) are seen as `audit_tenant_<tenantId>/cafAuditEvent/applicationAuditEvent`.
```
Again, a tenant does not make a call. 
audit_tenant_\<tenantId\> needs hanged to \<tenantId\>_audit.
Also could you explain some more what 'applicationAuditEvent' is here and give example if needed to help clarify this to the reader?

CJComac replied:
In the latest commit:
- Reworded to "On the audited application's first call to the Audit library, an index is created for the tenant. A tenant's Elasticsearch index, type and document meta-field identifiers (`index/type/doc`) are seen as `<tenantId>_audit/cafAuditEvent/applicationAuditEvent`."

- In the example where I show all of the events that belong to the "1_audit" index I have reworded it to "The above JSON, returned from Elasticsearch, illustrates all of the `applicationAuditEvent` documents belonging to the `1_audit` index. In other words each hit displayed is an audit event message belonging to a tenant application."

CJComac replied:
In the latest commit I have adjusted this again to illustrate how ES index and type meta-field identifiers are seen in the Audit solution with a table and a link to further reading on meta-field identifiers.

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,216 @@ title: Architecture
+        "max_score": 1,
+        "hits": [
+          {
+            "_index": "1_audit",
```
I'm wondering would it be best to give real example of a tenantid here instead of just 1. Same may also apply to other audit event metadata.

CJComac replied:
"On the audited application's first call to the Audit library, an index is created for the tenant. A tenant's Elasticsearch index, type and document meta-field identifiers (`index/type/doc`) are seen as `<tenantId>_audit/cafAuditEvent/applicationAuditEvent`." this to me describes the naming of indices. Do you think it would be better if I just replaced all instances of "1_audit" with "\<tenantId>_audit" ?

CJComac replied:
Dermot seemed happy with the data but I will get a more realistic tenant Id from him to use throughout the documentation

CJComac replied:
Updated tenant id from 1 to 00000001 in the latest commit

Commit:
Made Greg's feedback changes

Commit:
Made Connor's feedback changes

Commit:
Based on Connor's feedback:
-  Adjusted the description of how ES meta-field identifiers are seen in the Audit solution the use of a table.
- Moved descriptions of the returned JSON blobs to sit above the JSON blobs.

Commit:
Changed tenantId from 1 to 00000001

Commit:
Based on Connor's review I made adjustments to the Elasticsearch indexing section to give a better overview on tenant index and type mappings

mulhollandc commented on docs/pages/en-us/Architecture.md:
```diff
@@ -5,78 +5,233 @@ title: Architecture
+
+- `properties` : each of these are audit event parameter field names, present in all audit events, with the datatypes that each are stored as.
+- `dynamic_templates` : each of these are potential audit event parameter field names suffixed with a datatype identifier. An audit event parameter field name that matches a dynamic field is store with the corresponding datatype.
+
```
Should be "field is stored".

Commit:
Made minor grammar changes
